### PR TITLE
cpu/esp_common: allow configuration of UART0

### PR DESCRIPTION
**Pull request by gschorcht**
### Contribution description

This PR
- fixes the issue for ESP32 SoCs that UART0 signals can't be routed to arbitrary GPIOs and
- allows the configuration of the UART device used by the bootloader.

The UART interface and its configuration used by the STDIO are defined in RIOT using the define `STDIO_UART_DEV` and the configuration of the corresponding UART device in `periph_conf.h`. 

However, the bootloader compiled directly in ESP-IDF uses its own definitions `CONFIG_ESP_CONSOLE_UART_*` for the UART configuration. To be able to use a consistent UART configuration in RIOT and the bootloader, e.g. to see the output of the 2nd stage bootloader, these `CONFIG_ESP_CONSOLE_UART_*` can be defined via a set of KConfig variables in RIOT (not yet implemented in Kconfig):
- `CONSOLE_CONFIG_UART_NUM` defines the UART device to be used by the bootloader and by `STDIO_UART_DEV`
- `CONSOLE_CONFIG_UART_RX` and `CONSOLE_CONFIG_UART_TX` define the GPIOs to be used by the bootloader and should be the GPIOs as defined in `periph_conf.h` for the corresponding UART device.

### Testing procedure

Any ESP32 node should still work with `stdio_uart` and the default configuration. To test an alternative configuration, use
```
CFLAGS='-DUART1_TXD=5 -DUART1_RXD=4 -DCONFIG_CONSOLE_UART_NUM=1 -DCONFIG_CONSOLE_UART_TX=5 -DCONFIG_CONSOLE_UART_RX=4' USEMODULE=esp_log_startup BOARD=esp32-wroom-32 make -C tests/shell flash
```
The bootloader output and the STDIO should be routed to UART1 at GPIO4 and GPIO5.

### Issues/PRs references

Prerequisite for PR ##18863

### CODE REVIEW

File: cpu/esp32/bootloader/sdkconfig.h [hunk 1]
```diff
@@ -85,7 +85,25 @@ extern "C" {
  * Bootloader output baudrate, defined by the app settings as BAUD or
  * BOOTLOADER_BAUD.
  */
-#define CONFIG_ESP_CONSOLE_UART_BAUDRATE (RIOT_BOOTLOADER_BAUD)
+#define CONFIG_ESP_CONSOLE_UART_BAUDRATE    (RIOT_BOOTLOADER_BAUD)
+
+/*
```
Comment (benpicco) [inline cpu/esp32/bootloader/sdkconfig.h:90]:
```suggestion
/*
```

**Review by benpicco — COMMENTED** [review]
This breaks stdio on esp8266 :(

File: cpu/esp32/bootloader/sdkconfig.h (hunk 1 above)
Comment (gschorcht) [reply cpu/esp32/bootloader/sdkconfig.h:90]: I was afraid of that :worried: Let me check.

Comment (gschorcht) [reply cpu/esp32/bootloader/sdkconfig.h:90]: I'm not sure whether the reconfiguration of UART make sense at all for ESP8266. It has only one full functional UART interface. The TRM for ESP8266 is so pure, that I don't know whether the signals are hardwired or routable. Futhermore,  I don't know whether there is a ESP8266 board in universe that doesn't have connected the UART0 to the UART-to-USB bridge. So I would remove the configuration of GPIOs from `uart_init`.

**Comment by gschorcht** [comment]
> This breaks stdio on esp8266-esp-12x

I was afraid of that worried Let me check.

**Comment by gschorcht** [comment]
I'm not sure whether the reconfiguration of UART make sense at all for ESP8266. It has only one full functional UART interface. The TRM for ESP8266 is so pure, that I don't know whether the signals are hardwired or routable. Futhermore, I don't know whether there is a ESP8266 board in universe that doesn't have connected the UART0 to the UART-to-USB bridge. So I would remove the configuration of GPIOs from `uart_init`.

**Review by benpicco — APPROVED** [review]
Works now with default config on `esp32-wroom-32` and non-default config on `esp32s2-wemos-mini`.

Please squash

bors merge

**Comment by benpicco** [comment]
bors merge

**Comment by benpicco** [comment]
bors merge

**Comment by benpicco** [comment]
Huh, why close it?
Well, looks like bors merged it anyway.

**Pull request closed without merging**